create a README for the LocalizeMixin

margaree opened the pull request:
Much of this is copied from the d2l-localize-behavior, with mostly the usage and language resources being updated.

dlockhart commented on mixins/localize-mixin.md:
```diff
@@ -0,0 +1,294 @@
+				more: 'more',
+				less: 'less'
+			}
+		};
```
In this example, technically I don't think the language resources need to be in a private property right? Like they could have been defined in a constant somewhere in the file? Just thinking of a case where everyone copies this example and then with class inheritance you could have a base class and an inherited class both doing this and clobbering each other's language resources.

dlockhart replied:
Could we still use a constant defined outside the class (in the same file) instead of a local private property? That way I think we avoid redefining it for every instance of the class that gets created (less memory).

dlockhart commented on mixins/localize-mixin.md:
```diff
@@ -0,0 +1,294 @@
+		const namespace = `more-less:${lang}`;
+
+		if (proto.__localizationCache.requests[namespace]) {
+			return proto.__localizationCache.requests[namespace];
```
I don't think it's ideal for the sub-class to be calling into the base class's private properties here.

Since in this case (and the example below) it's checking a cache and if it's not there it's providing a way to fetch the value, maybe we could implement something in the base class that will do all that for us.

Then this class would just have to do something like this:

```javascript
async getLangResources(lang) {
    const namespace = `more-less:${lang}`;
    return this.localizationCache(namespace, () => {
        return this.__langResources[lang];
    });
}
```

And the dynamic example below:

```javascript
async getLangResources(lang) {
    const namespace = `more-less:${lang}`;
    return this.localizationCache(namespace, async () => {
        let translations;
        switch (lang) {
            case 'en':
                translations = await import('./locales/en.js');
                break;
            case 'ar':
                translations = await import('./locales/ar.js');
                break;
        }
        resolve(translations.val);
    });
}
```

So the underlying `localizationCache()` would:
- initialize the cache if it hasn't been already
- check if there's an entry for the key provided -- if so, return it
- otherwise it would call the function provided and put the result in the cache (which may be a promise)

margaree commented:
Changes made:
- Simplification of `getLangResources` implementation within components
- Components will define `namespaceBase` property value (e.g., `more-less`)
- Cache is completely handled in localize-mixin.js
- `this.__resources` was removed - because of this it became necessary to call `requestUpdate()` since __resources changing triggered render (let me know if I used this incorrectly)

dlockhart commented on mixins/localize-mixin.js:
```diff
@@ -54,35 +43,32 @@ export const LocalizeMixin = superclass => class extends superclass {
-						}
-						this._onRequestResponse(res, this.__language);
-					});
+				this.__namespace = `${this.__namespaceBase}:${this.__language}`;
```
Shouldn't this be `this.namespaceBase` (without the double underscores)?

margaree replied:
it definitely should

dlockhart commented on mixins/localize-mixin.md:
```diff
@@ -0,0 +1,265 @@
+	constructor() {
+		super();
+
+		this.namespaceBase = 'my-component';
```
Trying to think of ways it could avoid having to set this entirely. If only the mixin could know the element name (e.g. `d2l-more-less`), but I don't think it has access to that in Lit. In Polymer we had the `is` that was always set. Ah well.

dlockhart replied:
Maybe a static getter like how Lit does `properties` and `styles` might be better? I'd call it something like `localizeNamespace` (since `namespaceBase` it kind of generic and many mixins can be used on the same element).

margaree commented:
Recent changes:
- removed the cache from `LocalizeMixin` (it wasn't really providing any performance improvements)
- removed the need for both `getLanguage` and `getLangResources`
- renamed `getLangResources` to `getLocalizeResources` and made it static

I haven't updated the localize mixin readme yet to reflect these changes in case there is more churn

margaree commented on mixins/localize-mixin.js:
```diff
@@ -48,21 +37,15 @@ export const LocalizeMixin = superclass => class extends superclass {
-						this._onRequestResponse(res, this.__language);
+						this.__language = Object.keys(res)[0];
+						this.__resources = {};
+						this.__resources[this.__language] = res[this.__language];
```
^ the way I did this was weird. will change.

dlockhart commented on components/more-less/more-less.js:
```diff
@@ -107,6 +102,23 @@ class MoreLess extends LocalizeMixin(LitElement)  {
+			if (langResources[langs[i]]) {
+				return {
+					[langs[i]]: langResources[langs[i]]
+				};
```
Does it need to return a hash -- could it just return `langResources[langs[i]]`?

margaree replied:
I think we need the language though for `IntlMessageFormat` (https://github.com/BrightspaceUI/core/blob/master/mixins/localize-mixin.js#L240)

dlockhart replied:
Ah I see. To make it more clear, maybe it could be something like this:

```
return {
    lang: langs[i],
    resources: langResources[langs[i]]
};
```

That way we could also ask them to return other stuff in the future if we need it.

margaree commented:
I'll fix the more-less and localize mixin related linting failures in this PR but leave the others

margaree commented:
Any additional feedback on this?

dlockhart commented:
@dbatiste Agreed. Worth throwing that into an Issue so we don't forget about it?

dbatiste commented:
Sounds good - thx Margaree... you beat me to it.  ;)

svanherk commented:
This looks great, lots of nice detail!
